### ISSUE TITLE
Add synthetic Select.exceptSymmetric[ALL](Select)

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#9631]
Set theory knows a symmetric difference between two sets:
https://en.wikipedia.org/wiki/Symmetric_difference

![image](https://user-images.githubusercontent.com/734593/70239074-e558dd80-176a-11ea-8c50-6bb525b6f7ed.png)

```
A ⊕ B = (A ∪ B) ∖ (A ∩ B) = (A ∖ B) ∪ (B ∖ A)
```

The latest probably being preferrable for performance. So:

```sql
A EXCEPT SYMMETRIC B = (A EXCEPT B) UNION (B EXCEPT A)
A EXCEPT SYMMETRIC ALL B = (A EXCEPT ALL B) UNION ALL (B EXCEPT ALL A)
```

**Comment by knutwannheden** [comment]
I think this would be a really useful addition.

Often when I write a query like this I also want to know which of the tuples are only in `A` or `B`, so I end up with something like:

```sql
SELECT 'A', A.* FROM (A EXCEPT B) A
UNION
SELECT 'B', B.* FROM (B EXCEPT A) B
```

Any proposal how `EXCEPT SYMMETRIC` could be extended to support this as well?

If both the _identifying column_'s name and value is hardcoded, a plain keyword clause would suffice. But I think at least the column name should be configurable (to avoid any clashes), so maybe something like this (quite a mouthful 😟):

```sql
SELECT * FROM A
EXCEPT SYMMETRIC WITH INDICATOR MY_COL
SELECT * FROM B
```
Where the result would have an additional `MY_COL` column with hardcoded values (e.g. `1` and `2`). (Both `WITH` and `INDICATOR` are already reserved words in SQL.)

Note: Some dialects (e.g. HSQLDB) support the `CORRESPONDING` and even `CORRESPONDING BY` clause for the set operators, but that wouldn't really help with this, because the effect is that only the specified columns end up being returned.

**Comment by lukaseder** [comment]
> Any proposal how `EXCEPT SYMMETRIC` could be extended to support this as well?

I understand that some synthetic discriminator column could be useful with any kind of union, having done my own fair share of these in the past, not just with "symmetric differences", but with ordinary unions as well. However, this is a completely different kind of operator, that just happens to look very similar in the case of there being two operands.

Once we have more than two operands, it appears less useful to have such an "indicator" or "discriminator" column when looking at the venn diagram:

![image](https://user-images.githubusercontent.com/734593/70375346-09e4bf00-18fd-11ea-9357-ee2b56093af3.png)

And specifically at the formal explanation of how a symmetric difference of three sets will come to be:

![image](https://user-images.githubusercontent.com/734593/70375414-cc346600-18fd-11ea-8db0-33bcf46b4c8c.png)

To me, it appears that this indicator will be pretty useless, or at least quite difficult to use when 3 sets are involved.

I'm not opposed to the idea, but I think the approach might be the wrong one... In any case, we can support the vanilla operator and think about discriminators later.

One alternative approach is this:

```sql
SELECT *
FROM (
  SELECT 'A' discriminator, a.* FROM a
) AS a 
FULL JOIN (
  SELECT 'B' discriminator, b.* FROM b
) AS b
USING (a.col1, a.col2, ..., a.colN)
WHERE a.discriminator IS NULL
OR b.discriminator is NULL
```

> Note: Some dialects (e.g. HSQLDB) support the `CORRESPONDING` and even `CORRESPONDING BY` clause for the set operators, but that wouldn't really help with this, because the effect is that only the specified columns end up being returned.

I was not aware they supported this. Very nice. We should add support in jOOQ as well. Should be relatively easy to emulate

**Comment by knutwannheden** [comment]
> I understand that some synthetic discriminator column could be useful with any kind of union, having done my own fair share of these in the past, not just with "symmetric differences", but with ordinary unions as well.

Indeed. For `UNION ALL` this is of course trivial to support, but for a plain `UNION` it would be more involved, I think.

> To me, it appears that this indicator will be pretty useless, or at least quite difficult to use when 3 sets are involved.

Agreed. So far I have however never had the need for this and I think this would be a reasonable limitation.

> I'm not opposed to the idea, but I think the approach might be the wrong one... In any case, we can support the vanilla operator and think about discriminators later.

Sure, makes sense.

Regarding your alternative implementations using a `FULL JOIN`: The `SELECT *` would of course have to be written as `SELECT nvl(a.col1, b.col1) col1, nvl(a.col2, b.col2) col2, ...` (or some equivalent like `CASE`).

> I was not aware they supported this. Very nice. We should add support in jOOQ as well. Should be relatively easy to emulate

At least the `CORRESPONDING BY` shouldn't be hard. The `CORRESPONDING` clause would in the presence of an asterisk require us to access the table metadata.

**Comment by lukaseder** [comment]
> Agreed. So far I have however never had the need for this and I think this would be a reasonable limitation.

While I definitely see the use-case, we shouldn't jump to an implementation of this extension too quickly. I see a bunch of *very unreasonable limitations* in this area.

Allowing for some generalisation, what you're suggesting here is in fact an extension where we can project arbitrary expressions from set operation subqueries, which are not to be taken into consideration for the set operation itself, but still be projected. Perhaps as an extension to `CORRESPONDING BY`:

```sql
SELECT * FROM a
UNION 
  CORRESPONDING BY (x, y)
  PROJECTING 
    WHEN FIRST THEN 'a' AS discriminator
    WHEN SECOND THEN 'b' AS discriminator
  END
SELECT * FROM b
```

(I don't like the idea nor the syntax (let's not discuss that particular syntax), but this is what's really being suggested here, IMO)

In the absence of `ALL`, we would need to have some `ORDER BY` depending semantics. In a way, this works similar to `ON` in PostgreSQL's `DISTINCT ON`, which I find *very* confusing. Also, while the semantics of such a projection might be well defined with `EXCEPT`, we already have trouble using it with `UNION` and `INTERSECT`, as it is not exactly clear what to do when there's an ambiguity between subqueries' projections.

All in all, this operator extension raises *a lot* of questions while producing only marginal value, as users can easily write the very specific, and rarely useful SQL manually themselves.

> Regarding your alternative implementations using a `FULL JOIN`: The `SELECT *` would of course have to be written as `SELECT nvl(a.col1, b.col1) col1, nvl(a.col2, b.col2) col2, ...` (or some equivalent like `CASE`).

Nope, the beauty of `USING` or `NATURAL` is that you don't have to do that. Try it (in PostgreSQL) :-)

```sql
WITH 
  a(col1, col2) AS (VALUES (1, 1), (2, 2)),
  b(col1, col2) AS (VALUES (1, 2), (2, 2))
SELECT *
FROM (
  SELECT 'A' discriminator, a.* FROM a
) AS a 
FULL JOIN (
  SELECT 'B' discriminator, b.* FROM b
) AS b
USING (col1, col2)
WHERE a.discriminator IS NULL
OR b.discriminator is NULL;

WITH 
  a(col1, col2) AS (VALUES (1, 1), (2, 2)),
  b(col1, col2) AS (VALUES (1, 2), (2, 2))
SELECT *
FROM (
  SELECT 'A' discriminator_a, a.* FROM a
) AS a 
NATURAL FULL JOIN (
  SELECT 'B' discriminator_b, b.* FROM b
) AS b
WHERE a.discriminator_a IS NULL
OR b.discriminator_b is NULL;
```

> At least the `CORRESPONDING BY` shouldn't be hard. The `CORRESPONDING` clause would in the presence of an asterisk require us to access the table metadata.

Yes. Which also isn't hard. We have `Select.getSelect()` for this.

I forgot to link the existing feature request for `CORRESPONDING`: https://github.com/jOOQ/jOOQ/issues/5285
